src/Compilers/CSharp/Portable/Parser/Lexer_RawStringLiteral.cs

```diff
@@ -0,0 +1,347 @@
+                if (IsAtEndOfText(currentChar))
+                {
+                    this.AddError(TextWindow.Position, width: 0, ErrorCode.ERR_Unterminated_raw_string_literal);
+                    return false;
+                }
+
+                if (SyntaxFacts.IsNewLine(currentChar))
+                    return true;
+
+                if (currentChar == '"')
+                {
+                    // Don't allow a content line to contain a quote sequence that looks like a delimiter (or longer)
+                    currentQuoteCount = ConsumeQuoteSequence();
+                    if (currentQuoteCount >= startingQuoteCount)
+                    {
+                        this.AddError(
+                            position: TextWindow.Position - currentQuoteCount,
+                            width: currentQuoteCount,
+                            ErrorCode.ERR_Raw_string_literal_delimeter_must_be_on_its_own_line);
+                        return false;
+                    }
+                }
+                else
+                {
+                    TextWindow.AdvanceChar();
+                }
+            }
+        }
+
+        private void AddMultiLineRawStringLiteralLineContents(
+            StringBuilder indentationWhitespace,
+            StringBuilder currentLineWhitespace,
+            bool firstContentLine)
+        {
+            Debug.Assert(SyntaxFacts.IsNewLine(TextWindow.PeekChar()));
+
+            var newLineWidth = GetNewLineWidth(TextWindow.PeekChar());
+            for (var i = 0; i < newLineWidth; i++)
+            {
+                // the initial newline in `"""   \r\n` is not added to the contents.
+                if (!firstContentLine)
+                    _builder.Append(TextWindow.PeekChar());
+
+                TextWindow.AdvanceChar();
+            }
+
+            var lineStartPosition = TextWindow.Position;
+            currentLineWhitespace.Clear();
+            ConsumeWhitespace(currentLineWhitespace);
+
+            if (SyntaxFacts.IsNewLine(TextWindow.PeekChar()))
+            {
+                // a whitespace-only content line.  The indentation whitespace must be a prefix of the current line whitespace,
+                // or vice versa.  It is an error otherwise.
+                if (!StartsWith(indentationWhitespace, currentLineWhitespace) &&
+                    !StartsWith(currentLineWhitespace, indentationWhitespace))
+                {
+                    this.AddError(
+                        lineStartPosition,
+                        width: TextWindow.Position - lineStartPosition,
+                        ErrorCode.ERR_Line_does_not_start_with_the_same_whitespace_as_the_last_line_of_the_raw_string_literal);
+                    return;
+                }
+            }
+            else
+            {
+                // a content line with non-whitespace.  The indentation whitespace must be a prefix of the current line
+                // whitespace.  It is an error otherwise.
```

Comment:
Thanks much for the comments throughout the code. Definitely made the review easier